File: compilers/java/src/main/java/runar/compiler/passes/BuiltinArity.java
package runar.compiler.passes;

import java.util.List;
import java.util.Map;

/**
 * Builtin call arities, for validating ANF IR that never went through the
 * frontend (R-128 / R-165 family).
 *
 * <p>The source pipeline type-checks every call before codegen. {@code --ir}
 * runs no frontend at all, so a call with the wrong number of arguments used to
 * reach stack lowering, where each dispatch family consumes {@code args.size()}
 * from the stack MODEL and then emits a FIXED-arity opcode blob. Measured
 * through the six {@code --ir} CLIs on one file:
 *
 * <pre>
 *   cat(1 arg, needs 2)   go / ruby / rust / python / java compiled it to `7e`,
 *                         a bare OP_CAT with nothing beneath it
 *   assert(0 args)        the same five emitted an EMPTY script — the only
 *                         guard in the contract vanished, which is
 *                         anyone-can-spend, not merely wrong
 *                         (zig refused both, with InvalidBuiltin)
 * </pre>
 *
 * <p>This tier has no central builtin signature table to derive from — its
 * type-checker validates each family inline — so the counts are transcribed
 * from {@code compilers/go/ir/builtin_arity.go}, which is itself checked
 * against that tier's signature map by a drift test. The cross-tier gate is
 * {@code conformance/negatives/ir/} : every {@code I*.ir.json} must be refused
 * by all six tiers with an {@code --ir} mode, so a table that drifts here shows
 * up as an accepted negative rather than as silence.
 *
 * <p>Values are the ALLOWED argument counts. Two builtins accept more than one:
 * {@code assert} takes 1 or 2 (the optional message) and
 * {@code extractPrevOutputScript} takes 2 or 3 (the optional prefix length).
 * {@code merkleRootPoseidon2KB} is variadic by a rule and is handled in
 * {@link #check}.
 */
public final class BuiltinArity {

    private BuiltinArity() {}

    private static final Map<String, List<Integer>> ARITY = Map.ofEntries(
        Map.entry("abs", List.of(1)),
        Map.entry("assert", List.of(1, 2)),
        Map.entry("assertGroth16WitnessAssisted", List.of(0)),
        Map.entry("assertGroth16WitnessAssistedWithMSM", List.of(0)),
        Map.entry("bbExt4Inv0", List.of(4)),
        Map.entry("bbExt4Inv1", List.of(4)),
        Map.entry("bbExt4Inv2", List.of(4)),
        Map.entry("bbExt4Inv3", List.of(4)),
        Map.entry("bbExt4Mul0", List.of(8)),
        Map.entry("bbExt4Mul1", List.of(8)),
        Map.entry("bbExt4Mul2", List.of(8)),
        Map.entry("bbExt4Mul3", List.of(8)),
        Map.entry("bbFieldAdd", List.of(2)),
        Map.entry("bbFieldInv", List.of(1)),
        Map.entry("bbFieldMul", List.of(2)),
        Map.entry("bbFieldSub", List.of(2)),
        Map.entry("bin2num", List.of(1)),
        Map.entry("blake3Compress", List.of(2)),
        Map.entry("blake3Hash", List.of(1)),
        Map.entry("bn254FieldAdd", List.of(2)),
        Map.entry("bn254FieldInv", List.of(1)),
        Map.entry("bn254FieldMul", List.of(2)),
        Map.entry("bn254FieldNeg", List.of(1)),
        Map.entry("bn254FieldSub", List.of(2)),
        Map.entry("bn254G1Add", List.of(2)),
        Map.entry("bn254G1Negate", List.of(1)),
        Map.entry("bn254G1OnCurve", List.of(1)),
        Map.entry("bn254G1ScalarMul", List.of(2)),
        Map.entry("bn254MultiPairing3", List.of(27)),
        Map.entry("bn254MultiPairing4", List.of(20)),
        Map.entry("bn254Pairing", List.of(5)),
        Map.entry("bool", List.of(1)),
        Map.entry("buildChangeOutput", List.of(2)),
        Map.entry("cat", List.of(2)),
        Map.entry("checkMultiSig", List.of(2)),
        Map.entry("checkPreimage", List.of(1)),
        Map.entry("checkSig", List.of(2)),
        Map.entry("clamp", List.of(3)),
        Map.entry("currentBlockHeight", List.of(0)),
        Map.entry("divmod", List.of(2)),
        Map.entry("ecAdd", List.of(2)),
        Map.entry("ecEncodeCompressed", List.of(1)),
        Map.entry("ecMakePoint", List.of(2)),
        Map.entry("ecModReduce", List.of(2)),
        Map.entry("ecMul", List.of(2)),
        Map.entry("ecMulGen", List.of(1)),
        Map.entry("ecNegate", List.of(1)),
        Map.entry("ecOnCurve", List.of(1)),
        Map.entry("ecPointX", List.of(1)),
        Map.entry("ecPointY", List.of(1)),
        Map.entry("exit", List.of(1)),
        Map.entry("extractAmount", List.of(1)),
        Map.entry("extractHashPrevouts", List.of(1)),
        Map.entry("extractHashSequence", List.of(1)),
        Map.entry("extractInputIndex", List.of(1)),
        Map.entry("extractLocktime", List.of(1)),
        Map.entry("extractOutpoint", List.of(1)),
        Map.entry("extractOutputHash", List.of(1)),
        Map.entry("extractOutputs", List.of(1)),
        Map.entry("extractPrevOutputScript", List.of(2, 3)),
        Map.entry("extractScriptCode", List.of(1)),
        Map.entry("extractSequence", List.of(1)),
        Map.entry("extractSigHashType", List.of(1)),
        Map.entry("extractVersion", List.of(1)),
        Map.entry("gcd", List.of(2)),
        Map.entry("groth16PublicInput", List.of(1)),
        Map.entry("hash160", List.of(1)),
        Map.entry("hash256", List.of(1)),
        Map.entry("int2str", List.of(2)),
        Map.entry("kbExt4Inv0", List.of(4)),
        Map.entry("kbExt4Inv1", List.of(4)),
        Map.entry("kbExt4Inv2", List.of(4)),
        Map.entry("kbExt4Inv3", List.of(4)),
        Map.entry("kbExt4Mul0", List.of(8)),
        Map.entry("kbExt4Mul1", List.of(8)),
        Map.entry("kbExt4Mul2", List.of(8)),
        Map.entry("kbExt4Mul3", List.of(8)),
        Map.entry("kbFieldAdd", List.of(2)),
        Map.entry("kbFieldInv", List.of(1)),
        Map.entry("kbFieldMul", List.of(2)),
        Map.entry("kbFieldSub", List.of(2)),
        Map.entry("left", List.of(2)),
        Map.entry("len", List.of(1)),
        Map.entry("log2", List.of(1)),
        Map.entry("max", List.of(2)),
        Map.entry("merkleRootHash256", List.of(4)),
        Map.entry("merkleRootSha256", List.of(4)),
        Map.entry("min", List.of(2)),
        Map.entry("mulDiv", List.of(3)),
        Map.entry("num2bin", List.of(2)),
        Map.entry("p256Add", List.of(2)),
        Map.entry("p256EncodeCompressed", List.of(1)),
        Map.entry("p256Mul", List.of(2)),
        Map.entry("p256MulGen", List.of(1)),
        Map.entry("p256Negate", List.of(1)),
        Map.entry("p256OnCurve", List.of(1)),
        Map.entry("p384Add", List.of(2)),
        Map.entry("p384EncodeCompressed", List.of(1)),
        Map.entry("p384Mul", List.of(2)),
        Map.entry("p384MulGen", List.of(1)),
        Map.entry("p384Negate", List.of(1)),
        Map.entry("p384OnCurve", List.of(1)),
        Map.entry("pack", List.of(1)),
        Map.entry("percentOf", List.of(2)),
        Map.entry("pow", List.of(2)),
        Map.entry("requireOutputP2PKH", List.of(3)),
        Map.entry("reverseBytes", List.of(1)),
        Map.entry("right", List.of(2)),
        Map.entry("ripemd160", List.of(1)),
        Map.entry("safediv", List.of(2)),
        Map.entry("safemod", List.of(2)),
        Map.entry("sha256", List.of(1)),
        Map.entry("sha256Compress", List.of(2)),
        Map.entry("sha256Finalize", List.of(3)),
        Map.entry("sign", List.of(1)),
        Map.entry("split", List.of(2)),
        Map.entry("sqrt", List.of(1)),
        Map.entry("substr", List.of(3)),
        Map.entry("toByteString", List.of(1)),
        Map.entry("unpack", List.of(1)),
        Map.entry("verifyECDSA_P256", List.of(3)),
        Map.entry("verifyECDSA_P384", List.of(3)),
        Map.entry("verifyRabinSig", List.of(4)),
        Map.entry("verifySLHDSA_SHA2_128f", List.of(3)),
        Map.entry("verifySLHDSA_SHA2_128s", List.of(3)),
        Map.entry("verifySLHDSA_SHA2_192f", List.of(3)),
        Map.entry("verifySLHDSA_SHA2_192s", List.of(3)),
        Map.entry("verifySLHDSA_SHA2_256f", List.of(3)),
        Map.entry("verifySLHDSA_SHA2_256s", List.of(3)),
        Map.entry("verifySP1FRI", List.of(3)),
        Map.entry("verifyWOTS", List.of(3)),
        Map.entry("within", List.of(3)),
        Map.entry("__never__", List.of(0))
    );

    /**
     * Returns a diagnostic when {@code func} is a known builtin called with an
     * argument count it does not accept, or {@code null} when the call is fine
     * (or the name is not a builtin this table knows).
     */
    public static String check(String methodName, String bindingName, String func, int got) {
        if (func == null || func.isEmpty()) {
            return null;
        }
        if (func.equals("merkleRootPoseidon2KB")) {
            // 8 leaf elements + 8 per proof level + index + depth.
            if (got < 10) {
                return diagnostic(methodName, bindingName, func, got,
                    "at least 10 arguments (8 leaf + index + depth)");
            }
            if ((got - 10) % 8 != 0) {
                return diagnostic(methodName, bindingName, func, got, "8*depth + 10 arguments");
            }
            return null;
        }
        List<Integer> allowed = ARITY.get(func);
        if (allowed == null || allowed.contains(got)) {
            return null;
        }
        StringBuilder wanted = new StringBuilder();
        for (int i = 0; i < allowed.size(); i++) {
            if (i > 0) {
                wanted.append(i == allowed.size() - 1 ? " or " : ", ");
            }
            wanted.append(allowed.get(i));
        }
        return diagnostic(methodName, bindingName, func, got, wanted.toString());
    }

    private static String diagnostic(
        String methodName, String bindingName, String func, int got, String wanted
    ) {
        return "IR validation: method " + methodName + " binding " + bindingName
            + " calls " + func + "() with " + got + " argument(s); it takes " + wanted;
    }
}

File: compilers/go/ir/builtin_arity.go
package ir

// Builtin call arities, for validating ANF IR that never went through the
// frontend (R-163 / R-165 / CL-BUG-133 / CL-BUG-135).
//
// The source pipeline type-checks every call before codegen. `--ir` runs no
// frontend at all, so a call with the wrong number of arguments used to reach
// stack lowering, where each dispatch family pops `len(args)` from the stack
// MODEL and then emits a FIXED-arity opcode blob. Measured through the six
// `--ir` CLIs on one file:
//
//	cat(1 arg, needs 2)   go / ruby / rust / python compiled it to `7e`,
//	                      a bare OP_CAT with nothing beneath it
//	assert(0 args)        the same four emitted an EMPTY script — the only
//	                      guard in the contract vanished, which is
//	                      anyone-can-spend, not merely wrong
//	                      (java and zig refused both)
//
// This table is the arity half of `frontend.builtinFunctions`, duplicated here
// because `ir` cannot import `frontend` (the dependency runs the other way).
// `TestBuiltinArityMatchesFrontendTable` in the frontend package fails the
// build if the two ever disagree, so the copy cannot drift.
//
// Values are the ALLOWED argument counts. Two builtins accept more than one:
// `assert` takes 1 or 2 (the optional message), and `extractPrevOutputScript`
// takes 2 or 3 (the optional prefix length) — both special-cased in
// `typecheck.checkCallArgs` for exactly the same reason.
var BuiltinArity = map[string][]int{
	"abs":                                 {1},
	"assert":                              {1, 2},
	"assertGroth16WitnessAssisted":        {0},
	"assertGroth16WitnessAssistedWithMSM": {0},
	"bbExt4Inv0":                          {4},
	"bbExt4Inv1":                          {4},
	"bbExt4Inv2":                          {4},
	"bbExt4Inv3":                          {4},
	"bbExt4Mul0":                          {8},
	"bbExt4Mul1":                          {8},
	"bbExt4Mul2":                          {8},
	"bbExt4Mul3":                          {8},
	"bbFieldAdd":                          {2},
	"bbFieldInv":                          {1},
	"bbFieldMul":                          {2},
	"bbFieldSub":                          {2},
	"bin2num":                             {1},
	"blake3Compress":                      {2},
	"blake3Hash":                          {1},
	"bn254FieldAdd":                       {2},
	"bn254FieldInv":                       {1},
	"bn254FieldMul":                       {2},
	"bn254FieldNeg":                       {1},
	"bn254FieldSub":                       {2},
	"bn254G1Add":                          {2},
	"bn254G1Negate":                       {1},
	"bn254G1OnCurve":                      {1},
	"bn254G1ScalarMul":                    {2},
	"bn254MultiPairing3":                  {27},
	"bn254MultiPairing4":                  {20},
	"bn254Pairing":                        {5},
	"bool":                                {1},
	"buildChangeOutput":                   {2},
	"cat":                                 {2},
	"checkMultiSig":                       {2},
	"checkPreimage":                       {1},
	"checkSig":                            {2},
	"clamp":                               {3},
	"currentBlockHeight":                  {0},
	"divmod":                              {2},
	"ecAdd":                               {2},
	"ecEncodeCompressed":                  {1},
	"ecMakePoint":                         {2},
	"ecModReduce":                         {2},
	"ecMul":                               {2},
	"ecMulGen":                            {1},
	"ecNegate":                            {1},
	"ecOnCurve":                           {1},
	"ecPointX":                            {1},
	"ecPointY":                            {1},
	"exit":                                {1},
	"extractAmount":                       {1},
	"extractHashPrevouts":                 {1},
	"extractHashSequence":                 {1},
	"extractInputIndex":                   {1},
	"extractLocktime":                     {1},
	"extractOutpoint":                     {1},
	"extractOutputHash":                   {1},
	"extractOutputs":                      {1},
	"extractPrevOutputScript":             {2, 3},
	"extractScriptCode":                   {1},
	"extractSequence":                     {1},
	"extractSigHashType":                  {1},
	"extractVersion":                      {1},
	"gcd":                                 {2},
	"groth16PublicInput":                  {1},
	"hash160":                             {1},
	"hash256":                             {1},
	"int2str":                             {2},
	"kbExt4Inv0":                          {4},
	"kbExt4Inv1":                          {4},
	"kbExt4Inv2":                          {4},
	"kbExt4Inv3":                          {4},
	"kbExt4Mul0":                          {8},
	"kbExt4Mul1":                          {8},
	"kbExt4Mul2":                          {8},
	"kbExt4Mul3":                          {8},
	"kbFieldAdd":                          {2},
	"kbFieldInv":                          {1},
	"kbFieldMul":                          {2},
	"kbFieldSub":                          {2},
	"left":                                {2},
	"len":                                 {1},
	"log2":                                {1},
	"max":                                 {2},
	"merkleRootHash256":                   {4},
	"merkleRootSha256":                    {4},
	"min":                                 {2},
	"mulDiv":                              {3},
	"num2bin":                             {2},
	"p256Add":                             {2},
	"p256EncodeCompressed":                {1},
	"p256Mul":                             {2},
	"p256MulGen":                          {1},
	"p256Negate":                          {1},
	"p256OnCurve":                         {1},
	"p384Add":                             {2},
	"p384EncodeCompressed":                {1},
	"p384Mul":                             {2},
	"p384MulGen":                          {1},
	"p384Negate":                          {1},
	"p384OnCurve":                         {1},
	"pack":                                {1},
	"percentOf":                           {2},
	"pow":                                 {2},
	"requireOutputP2PKH":                  {3},
	"reverseBytes":                        {1},
	"right":                               {2},
	"ripemd160":                           {1},
	"safediv":                             {2},
	"safemod":                             {2},
	"sha256":                              {1},
	"sha256Compress":                      {2},
	"sha256Finalize":                      {3},
	"sign":                                {1},
	"split":                               {2},
	"sqrt":                                {1},
	"substr":                              {3},
	"toByteString":                        {1},
	"unpack":                              {1},
	"verifyECDSA_P256":                    {3},
	"verifyECDSA_P384":                    {3},
	"verifyRabinSig":                      {4},
	"verifySLHDSA_SHA2_128f":              {3},
	"verifySLHDSA_SHA2_128s":              {3},
	"verifySLHDSA_SHA2_192f":              {3},
	"verifySLHDSA_SHA2_192s":              {3},
	"verifySLHDSA_SHA2_256f":              {3},
	"verifySLHDSA_SHA2_256s":              {3},
	"verifySP1FRI":                        {3},
	"verifyWOTS":                          {3},
	"within":                              {3},
}

// AllowedArity reports the argument counts a builtin accepts, and whether the
// name is a known FIXED-arity builtin at all.
func AllowedArity(name string) ([]int, bool) {
	counts, ok := BuiltinArity[name]
	return counts, ok
}

// VariadicArityOK checks the builtins whose arity is a RULE rather than a
// count. Returns `isVariadic=false` for every other name.
//
// `merkleRootPoseidon2KB` takes 8 leaf elements + 8 per proof level + index +
// depth, i.e. 8*depth + 10, which is why its signature-table entry carries no
// params at all and `typecheck.checkCallArgs` special-cases it. The same rule
// is applied here so the `--ir` path gets the real check rather than a pass.
func VariadicArityOK(name string, got int) (ok bool, rule string, isVariadic bool) {
	switch name {
	case "merkleRootPoseidon2KB":
		if got < 10 {
			return false, "at least 10 arguments (8 leaf + index + depth)", true
		}
		if (got-10)%8 != 0 {
			return false, "8*depth + 10 arguments", true
		}
		return true, "", true
	}
	return false, "", false
}
